### PR TITLE
Finetuning module, Hybrid module, New models

### PR DESCRIPTION
**Pull request by The-Mats**
This branch includes the changes for my Bachelor thesis and can be merged into main afterwards:
### Finetuning Module
- [x] Add Finetune Module
- [x] Add needed configs
- [x] Add run script

### Hybrid Module
- [x] Add Hybrid Module that inherits from Finetune Module
- [x] Add needed configs
- [x] Add run script

### Embedding module
- [x] Non-centered padding in Embedding Module like in Pipeline
- [x] Removed resampling from Embedding Module
- [x] Add run script

### Models
- Merged BEATs into branch
- Merged AudioMAE into branch
- [x] Add get_embedding() for ConvNeXt
- [x] Add get embedding() for EAT
- [x] Integrate AVES model 
- [x] Test with AVES
- [x] Integrate BioLingual model
- [x] Test with BioLingual 

### Other
- Reactivated valid accuracy metric
- Added ability to use "ckpt_path" for testing
- Added plot notebook and result CSVs

Maybe the results folder should be excluded in the merge 🤔 It contains the notebook for my plots, the results CSVs and the plot PNGs.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -29,6 +29,10 @@ dataset_/*
 
 
 birdset/output/*
+LTU.pth/*
```
Comment (raphaelschwinger) [inline .gitignore]: Move the model checkpoints into the `/models` folder. Then you also do not need to gitignore them

File: birdset/datamodule/embedding_datamodule.py [hunk 2]
```diff
@@ -251,21 +251,14 @@ def _get_embedding(self, audio):
...
     def _zero_pad(self, audio):
         desired_num_samples = self.max_length * self.sampling_rate 
         current_num_samples = audio.shape[0]
         padding = desired_num_samples - current_num_samples
         if padding > 0:
-            #print('padding')
-            pad_left = padding // 2
-            pad_right = padding - pad_left
-            audio = torch.nn.functional.pad(audio, (pad_left, pad_right))
+            audio = torch.nn.functional.pad(audio, (0, padding))
```
Comment (raphaelschwinger) [inline birdset/datamodule/embedding_datamodule.py:282]: top!

File: birdset/modules/finetune_module.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+from torch.optim import AdamW, Optimizer
+from birdset.configs import NetworkConfig, LoggingParamsConfig, LRSchedulerConfig, MulticlassMetricsConfig, MultilabelMetricsConfig, MultilabelMetricsConfig as MetricsConfig
+from birdset.datamodule.embedding_datamodule import EmbeddingModuleConfig
+from birdset.utils import pylogger
+
+import torch
+import math
+
+log = pylogger.get_pylogger(__name__)
+class FinetuneModule(BaseModule):
```
Comment (raphaelschwinger) [inline birdset/modules/finetune_module.py:15]: add pydoc

File: birdset/modules/finetune_module.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        )
+        self.embedding_model = embedding_model.model
+        self.sampling_rate = embedding_model.sampling_rate
+        self.max_length = embedding_model.length  
+        log.info(f"FINETUNING using embedding model:{embedding_model.model_name} (Sampling Rate:{self.sampling_rate}, Window Size:{self.max_length})")
+
+    def configure_optimizers(self):
+        self.optimizer = self.optimizer(list(self.model.parameters())+list(self.embedding_model.parameters())) #! Changed this
+        if self.lr_scheduler is not None:
+            # TODO: Handle the case when we do not want warmup
```
Comment (raphaelschwinger) [inline birdset/modules/finetune_module.py:83]: If you do not attend to solve those TODOs in this PR, please create an issue and add a link to the issue in the comments.

File: birdset/modules/finetune_module.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            num_training_steps = math.ceil((self.num_epochs * self.len_trainset) / self.batch_size * self.num_gpus)
+            num_warmup_steps = math.ceil(
+                    num_training_steps * self.lr_scheduler.warmup_ratio
+                )
+            # TODO: Handle the case when drop_last=True more explicitly
```
Comment (raphaelschwinger) [inline birdset/modules/finetune_module.py:88]: same

File: birdset/modules/hybrid_module.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+from birdset.datamodule.embedding_datamodule import EmbeddingModuleConfig
+from birdset.utils import pylogger
+from birdset.modules.finetune_module import FinetuneModule
+from torch.optim.lr_scheduler import LambdaLR
+
+import torch
+import math
+
+log = pylogger.get_pylogger(__name__)
+class HybridModule(FinetuneModule):
```
Comment (raphaelschwinger) [inline birdset/modules/hybrid_module.py:14]: add pydoc

File: birdset/modules/hybrid_module.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                # Reduce LR to zero after warmup
+                return 0.0
+            else:
+                # Gradually increase LR from 0 to max_lr after unfreezing
+                total_epochs = self.trainer.max_epochs
+                progress = (epoch - self.ft_max_epochs+1) / (total_epochs - self.ft_max_epochs+1)
+                return progress * (self.ft_lr / 1e-2)
+        
+        scheduler = LambdaLR(optimizer, lr_lambda=lr_lambda)
+        return [optimizer], [scheduler]'''
```
Comment (raphaelschwinger) [inline birdset/modules/hybrid_module.py]: remove all commented out code, if you think its worth saving, put it into another function and add a config parameter to be able to use it.

File: birdset/modules/models/LTU.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+from torch import nn
+from ...utils.label_utils import get_label_to_class_mapping_from_metadata
+
+from .embedding_abstract import EmbeddingModel
+
+class LTU(nn.Module):
+    """
+    
+    Attributes:
+        PERCH_TF_HUB_URL (str): URL to the TensorFlow Hub model for bird vocalization.
```
Comment (raphaelschwinger) [inline birdset/modules/models/LTU.py]: Fix docs, replace perch with LTU.

File: birdset/modules/models/aves.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
+import torch
+import datasets
+import torch.nn as nn
+from typing import Tuple
+from birdset.configs import PretrainInfoConfig
+from torchaudio.models import wav2vec2_model
+import json
+from typing import Optional
+
+class AvesClassifier(nn.Module):
```
Comment (raphaelschwinger) [inline birdset/modules/models/aves.py:12]: add docs, including original paper and code references

**Review by raphaelschwinger — CHANGES_REQUESTED** [review]
Hi @XgamerTV,
I looked through your code, looks good! Lets merge it into main soon. To do so, please address the comments and:
- Merge main.
- Haven't we added BEATs in another PR? Remove it here.
- For all classes, make sure there are appropriate docs, especially for the model classes, add links to the paper, the code repository and there licence.
- Remove all usused code (commented out code). If you want to keep it, make it usable through hydra.
- Describe in the README what experiments you added and how to run them.

File: .gitignore (hunk 1 above)
Comment (The-Mats) [reply .gitignore]: This is something old I think, the checkpoints are all in the folder

**Comment by The-Mats** [comment]
Should I move the experiment configs in some special folder like "Thesis" or into Biofoundation? They are in local right now

**Comment by The-Mats** [comment]
Ah I see some experiments were renamed automatically with the merge but others are still in the local folder as well as all the finetune and hybrid experiments

File: birdset/modules/finetune_module.py (hunk 4 above)
Comment (The-Mats) [reply birdset/modules/finetune_module.py:83]: The TODOs come straight from the BaseModule 🤔

File: birdset/modules/models/LTU.py (hunk 8 above)
Comment (The-Mats) [reply birdset/modules/models/LTU.py]: I didn't use this model I think it may have come from merging BEATS into this branch

**Comment by The-Mats** [comment]
I made small changes to BEATS for example that the feature extractor is only frozen if train_classifier is true so that finetuning works.

**Comment by The-Mats** [comment]
I added two plots to the README for reference which could be removed or changed in the future when the BioFoundation Paper is done :)
